File: src/electron/etl/screentime.ts
import fs from "fs";
import path from "path";
import os from "os";
import sqlite3, { Database } from "better-sqlite3";

export const screenTimeSql  = `
CREATE TABLE IF NOT EXISTS screen_time (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  app TEXT NOT NULL,
  device_id TEXT,
  device_model TEXT,
  usage REAL,
  timezone INTEGER,
  created_at TEXT,
  start_time TEXT,
  end_time TEXT,
  UNIQUE(app, device_id, start_time)
)
`
export interface ScreenTimeRecord {
  app: string;
  device_id: string;
  device_model: string;
  usage: number;
  timezone: number | null;
  created_at: string;
  start_time: string;
  end_time: string;
}

export function etlScreentime(db: Database) {
  const rows = extractScreenTime();
  const records = transformScreenTime(rows);
  loadScreenTimeToSqlite(db, records)
}

interface ScreenTimeRaw {
  app: string;
  usage: number;
  start_time: number;
  end_time: number;
  created_at: number;
  tz: number | null;
  device_id: string | null;
  device_model: string | null;
}


const knowledgeDb = path.join(
  os.homedir(),
  "Library/Application Support/Knowledge/knowledgeC.db"
);

function extractScreenTime(): ScreenTimeRaw[] {
  if (!fs.existsSync(knowledgeDb)) {
    throw new Error(`Could not find knowledgeC.db at ${knowledgeDb}.`);
  }

  try {
    fs.accessSync(knowledgeDb, fs.constants.R_OK);
  } catch (err) {
    throw new Error(
      `The knowledgeC.db at ${knowledgeDb} is not readable. Please grant full disk access to the application running the script (Terminal, iTerm, VSCode, etc.).`
    );
  }

  const db = sqlite3(knowledgeDb, { readonly: true });

  const query = `
    SELECT
      ZOBJECT.ZVALUESTRING AS app, 
      (ZOBJECT.ZENDDATE - ZOBJECT.ZSTARTDATE) AS usage,
      (ZOBJECT.ZSTARTDATE + 978307200) as start_time, 
      (ZOBJECT.ZENDDATE + 978307200) as end_time,
      (ZOBJECT.ZCREATIONDATE + 978307200) as created_at, 
      ZOBJECT.ZSECONDSFROMGMT AS tz,
      ZSOURCE.ZDEVICEID AS device_id,
      ZMODEL AS device_model
    FROM
      ZOBJECT 
      LEFT JOIN ZSTRUCTUREDMETADATA ON ZOBJECT.ZSTRUCTUREDMETADATA = ZSTRUCTUREDMETADATA.Z_PK 
      LEFT JOIN ZSOURCE ON ZOBJECT.ZSOURCE = ZSOURCE.Z_PK 
      LEFT JOIN ZSYNCPEER ON ZSOURCE.ZDEVICEID = ZSYNCPEER.ZDEVICEID
    WHERE
      ZSTREAMNAME = '/app/usage'
    ORDER BY
      ZSTARTDATE DESC
  `;

  const rows: ScreenTimeRaw[] = db.prepare(query).all() as ScreenTimeRaw[];
  db.close();

  return rows;
}

function transformScreenTime(rows: ScreenTimeRaw[]): ScreenTimeRecord[] {
  return rows.map((r) => {
    const { app, usage, start_time, end_time, created_at, tz, device_id, device_model } = r;
    return {
      app,
      device_id: device_id ?? "Unknown",
      device_model: device_model ?? "Unknown",
      usage,
      timezone: tz,
      created_at: new Date(created_at * 1000).toISOString(),
      start_time: new Date(start_time * 1000).toISOString(),
      end_time: new Date(end_time * 1000).toISOString(),
    };
  });
}

function loadScreenTimeToSqlite(db: Database, records: ScreenTimeRecord[]) {
  const insert = db.prepare(`
    INSERT OR IGNORE INTO screen_time
      (app, device_id, device_model, usage, timezone, created_at, start_time, end_time)
    VALUES
      (@app, @device_id, @device_model, @usage, @timezone, @created_at, @start_time, @end_time)
  `);

  const insertMany = db.transaction((records: ScreenTimeRecord[]) => {
    for (const record of records) insert.run(record);
  });

  insertMany(records);
  console.log(`Inserted ${records.length} screen time records.`);
}

File: pipeline/extractors/screentime.py
"""
macOS Screen Time extractor.

Reads app usage directly from the macOS knowledgeC.db SQLite database and
uploads aggregated daily totals to R2.

Requires Full Disk Access for the app running the script
(Terminal, iTerm, VSCode, etc.) — grant it in:
  System Settings → Privacy & Security → Full Disk Access

Run via scripts/sync_screentime.py on a cron schedule.
"""

from __future__ import annotations

import os
import sqlite3
from datetime import datetime, timezone
from pathlib import Path

import polars as pl

from pipeline import r2 as R2
from pipeline.r2 import R2Client

METRIC = "app_usage"
UNIT = "minutes"
LABEL = "Screen time"

_DB = Path.home() / "Library/Application Support/Knowledge/knowledgeC.db"

# Apple's CoreData epoch starts 2001-01-01 — offset to Unix epoch
_APPLE_EPOCH_OFFSET = 978_307_200

_QUERY = """
SELECT
    ZOBJECT.ZVALUESTRING                        AS app,
    (ZOBJECT.ZENDDATE - ZOBJECT.ZSTARTDATE)     AS usage_secs,
    (ZOBJECT.ZSTARTDATE + 978307200)            AS start_unix,
    ZOBJECT.ZSECONDSFROMGMT                     AS tz_offset
FROM
    ZOBJECT
    LEFT JOIN ZSOURCE ON ZOBJECT.ZSOURCE = ZSOURCE.Z_PK
WHERE
    ZSTREAMNAME = '/app/usage'
    AND ZOBJECT.ZVALUESTRING IS NOT NULL
    AND (ZOBJECT.ZENDDATE - ZOBJECT.ZSTARTDATE) > 0
ORDER BY
    ZOBJECT.ZSTARTDATE DESC
"""


def run(r2: R2Client) -> None:
    rows = _query_db()
    if not rows:
        print("[screentime] no data returned, skipping")
        return

    df = _to_df(rows)
    print(f"[screentime/{METRIC}] {len(df)} rows")
    R2.store_partitions(r2, "screentime", METRIC, df)
    R2.write_web_json(r2, "screentime", METRIC, UNIT, LABEL)


def _query_db(db_path: Path = _DB) -> list[tuple]:
    if not db_path.exists():
        raise FileNotFoundError(f"knowledgeC.db not found at {db_path}")
    if not os.access(db_path, os.R_OK):
        raise PermissionError(
            f"{db_path} is not readable. "
            "Grant Full Disk Access to your terminal in "
            "System Settings → Privacy & Security → Full Disk Access"
        )
    with sqlite3.connect(f"file:{db_path}?mode=ro", uri=True) as con:
        return con.execute(_QUERY).fetchall()


def _to_df(rows: list[tuple]) -> pl.DataFrame:
    apps, usage_secs, start_unix, tz_offsets = zip(*rows)

    # Localise each timestamp using the per-record tz offset stored in the DB
    dates = [
        datetime.fromtimestamp(ts + (tz or 0), tz=timezone.utc).date()
        for ts, tz in zip(start_unix, tz_offsets)
    ]

    return (
        pl.DataFrame({
            "date":     list(dates),
            "value":    [round(s / 60, 1) for s in usage_secs],
            "category": list(apps),
        }, schema={"date": pl.Date, "value": pl.Float64, "category": pl.Utf8})
        .group_by(["date", "category"])
        .agg(pl.col("value").sum().round(1))
        .sort("date")
    )
